File: scripts/compress.js
#!/usr/bin/env node

import * as fs from "node:fs";
import * as path from "node:path";
import * as zlib from "node:zlib";

const usage = () => {
  console.error(`Usage: node compress <input.json> [--out <outputBasePath>]
  - Compresses JSON using Gzip, Brotli, and Deflate.
  - Outputs: <outputBasePath>.gz, .br, .deflate`);
  process.exit(1);
};

const getArgs = () => {
  const args = process.argv.slice(2);
  if (args.length === 0) usage();

  const inputPath = path.resolve(process.cwd(), args[0]);
  const outFlagIndex = args.indexOf("--out");
  const outputPath =
    outFlagIndex !== -1 && args[outFlagIndex + 1]
      ? path.resolve(process.cwd(), args[outFlagIndex + 1])
      : inputPath;

  return { path: { input: inputPath, output: outputPath } };
};

/**
 * @param {number} bytes
 */
const formatSize = (bytes) => `${(bytes / 1024).toFixed(2)} KiB`;

/**
 *
 * @param {unknown} jsonData
 * @param {string} basePath
 */
const compressJson = async (jsonData, basePath) => {
  const minified = JSON.stringify(jsonData);
  const buffer = Buffer.from(minified, "utf8");

  const compressions = {
    gzip: zlib.gzipSync(buffer, { level: zlib.constants.Z_BEST_COMPRESSION }),
    // deflate: zlib.deflateSync(buffer, { level: zlib.constants.Z_BEST_COMPRESSION }),
    brotli: zlib.brotliCompressSync(buffer, {
      params: { [zlib.constants.BROTLI_PARAM_QUALITY]: zlib.constants.BROTLI_MAX_QUALITY },
    }),
  };

  console.log(`ORIGINAL: ${formatSize(buffer.length)}`);

  await Promise.all(
    Object.entries(compressions).map(async ([type, data]) => {
      const ratio = ((data.length / buffer.length) * 100).toFixed(1);
      const outPath = `${basePath}.${type === "brotli" ? "br" : type}`;

      await fs.promises.writeFile(outPath, data);

      console.log(
        `${type.toUpperCase()}: ${formatSize(data.length)} (${ratio}% of original) → ${outPath}`,
      );
    }),
  );
};

const main = async () => {
  const { path } = getArgs();

  if (!fs.existsSync(path.input)) {
    console.error(`❌ Input file not found: ${path.input}`);
    process.exit(1);
  }

  try {
    const content = await fs.promises.readFile(path.input, "utf8");
    const json = JSON.parse(content);
    await compressJson(json, path.output);
  } catch (error) {
    // @ts-expect-error
    console.error(`❌ Error: ${error.message}`);
    process.exit(1);
  }
};

main();

File: scripts/compress.py
# /// script
# dependencies = ["brotli"]
# ///
"""Compress a JSON file with Gzip and Brotli.

Usage:
    uv run compress.py <input.json> [--out <outputBasePath>]

Outputs: <outputBasePath>.gz, <outputBasePath>.br
"""

import argparse
import gzip
import json
import sys
from pathlib import Path

import brotli


def format_size(num_bytes: int) -> str:
    return f"{num_bytes / 1024:.2f} KiB"


def compress_json(json_data: object, base_path: Path) -> None:
    minified = json.dumps(json_data, separators=(",", ":"))
    buffer = minified.encode("utf-8")

    compressions = {
        "gzip": gzip.compress(buffer, compresslevel=9),
        "brotli": brotli.compress(buffer, quality=11),
    }

    print(f"ORIGINAL: {format_size(len(buffer))}")

    for kind, data in compressions.items():
        ratio = (len(data) / len(buffer)) * 100
        extension = "br" if kind == "brotli" else kind
        out_path = base_path.with_name(f"{base_path.name}.{extension}")
        out_path.write_bytes(data)
        print(f"{kind.upper()}: {format_size(len(data))} ({ratio:.1f}% of original) -> {out_path}")


def main() -> None:
    parser = argparse.ArgumentParser(description="Compress JSON using Gzip and Brotli.")
    parser.add_argument("input", type=Path)
    parser.add_argument("--out", dest="output", type=Path, default=None)
    args = parser.parse_args()

    input_path = args.input.resolve()
    output_path = (args.output or args.input).resolve()

    if not input_path.exists():
        print(f"\u274c Input file not found: {input_path}", file=sys.stderr)
        sys.exit(1)

    try:
        content = input_path.read_text(encoding="utf-8")
        data = json.loads(content)
        compress_json(data, output_path)
    except Exception as error:  # noqa: BLE001 - mirrors the original's catch-all
        print(f"\u274c Error: {error}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
